presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,137 @@
+ * vary a lot based on the particular input page.
+ * <p>
+ * Considering the CPU time required to process(filter, project) a full (~1MB) page returned by a
+ * connector, the CPU cost of memory copying (< 50kb, < 1024 rows) is supposed to be negligible
+ */
+public class MergingPageOutput
+        implements Iterator<Page>
+{
+    private static final long DEFAULT_MIN_PAGE_SIZE_IN_BYTES = 50 * 1024;
+    private static final long DEFAULT_MIN_ROW_COUNT = 1024;
+
+    private final List<Type> types;
+    private final PageBuilder pageBuilder;
+    private final Queue<Page> outputQueue = new LinkedList<>();
+
+    private final long minPageSizeInBytes;
+    private final long minRowCount;
+
+    public MergingPageOutput(Iterable<? extends Type> types)
+    {
+        this(types, DEFAULT_MIN_PAGE_SIZE_IN_BYTES, DEFAULT_MIN_ROW_COUNT);
+    }
+
+    public MergingPageOutput(Iterable<? extends Type> types, long minPageSizeInBytes, long minRowCount)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.minPageSizeInBytes = minPageSizeInBytes;
+        this.minRowCount = minRowCount;
+        pageBuilder = new PageBuilder(this.types);
+    }
+
+    public void accept(Iterator<Page> pages)
+    {
+        requireNonNull(pages, "pages is null");
+        pages.forEachRemaining(this::accept);
+    }
+
+    public void accept(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        // avoid memory copying for pages that are big enough
+        if (page.getSizeInBytes() > minPageSizeInBytes || page.getPositionCount() > minRowCount) {
+            flush();
+            outputQueue.add(page);
+            return;
+        }
+
+        append(page);
+    }
+
+    public void flush()
+    {
+        if (!pageBuilder.isEmpty()) {
+            Page output = pageBuilder.build();
+            pageBuilder.reset();
+            outputQueue.add(output);
+        }
+    }
+
+    private void append(Page page)
+    {
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            if (pageBuilder.isFull()) {
+                flush();
+            }
+            pageBuilder.declarePosition();
+            for (int channel = 0; channel < types.size(); channel++) {
+                Type type = types.get(channel);
+                type.appendTo(page.getBlock(channel), position, pageBuilder.getBlockBuilder(channel));
+            }
+        }
+        if (pageBuilder.isFull()) {
+            flush();
+        }
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        return !outputQueue.isEmpty();
+    }
+
+    @Override
+    public Page next()
+    {
+        checkState(!outputQueue.isEmpty(), "outputQueue is empty");
+        return outputQueue.poll();
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        return pageBuilder.getRetainedSizeInBytes() + outputQueue.stream().mapToLong(Page::getRetainedSizeInBytes).sum();
```

Comment:
You already account for the dynamically sized fields (`outputQueue`, `pageBuilder`), what's missing here is the constant object overhead. We account for the object overhead in almost all our memory tracking code (`ClassLayout.parseClass(K.class).instanceSize()`), so it's good to be consistent.